libBlueSky/SizeGrip.cpp

```diff
@@ -0,0 +1,224 @@
+#include "SizeGrip.hpp"
+
+#include <QAbstractScrollArea>
+#include <QApplication>
+#include <QDesktopWidget>
+#include <QLayout>
+#include <QMouseEvent>
+
+namespace BlueSky {
+
+SizeGrip::SizeGrip(QWidget* parent)
+    : QSizeGrip(parent)
+    , mResizeDirections(ResizeDefault)
+    , mParentWindow(parent)
+    , mGotMousePress(false)
+    , dxMax(0)
+    , dyMax(0)
+    , mCorner(mParentWindow->isLeftToRight() ? Qt::BottomRightCorner : Qt::BottomLeftCorner)
+{
+    Q_ASSERT(parent);
+    resize(sizeHint());
+}
+
+SizeGrip::ResizeDirections SizeGrip::resizeBehaviour() const
+{
+    return mResizeDirections;
+}
+
+void SizeGrip::setResizeBehaviour(int behaviour)
+{
+    mResizeDirections = static_cast<ResizeDirections>(behaviour);
+}
+
+void SizeGrip::mousePressEvent(QMouseEvent* ev)
+{
+    if (ev->button() != Qt::LeftButton) {
+        QWidget::mousePressEvent(ev);
+        return;
+    }
+
+    QWidget *tlw = mParentWindow;
+    mGotMousePress = true;
+    mStartPos = ev->globalPos();
+    mDlgRect = mParentWindow->geometry();
+
+#ifdef Q_WS_X11
+    // Use a native X11 sizegrip for "real" top-level windows if supported.
+    if (tlw->isWindow() && X11->isSupportedByWM(ATOM(_NET_WM_MOVERESIZE))
+        && !(tlw->windowFlags() & Qt::X11BypassWindowManagerHint)
+        && !tlw->testAttribute(Qt::WA_DontShowOnScreen) && !qt_widget_private(tlw)->hasHeightForWidth()) {
+        XEvent xev;
+        xev.xclient.type = ClientMessage;
+        xev.xclient.message_type = ATOM(_NET_WM_MOVERESIZE);
+        xev.xclient.display = X11->display;
+        xev.xclient.window = tlw->winId();
+        xev.xclient.format = 32;
+        xev.xclient.data.l[0] = ev->globalPos().x();
+        xev.xclient.data.l[1] = ev->globalPos().y();
+        if (atBottom())
+            xev.xclient.data.l[2] = atLeft() ? 6 : 4; // bottomleft/bottomright
+        else
+            xev.xclient.data.l[2] = atLeft() ? 0 : 2; // topleft/topright
+        xev.xclient.data.l[3] = Button1;
+        xev.xclient.data.l[4] = 0;
+        XUngrabPointer(X11->display, X11->time);
+        XSendEvent(X11->display, QX11Info::appRootWindow(x11Info().screen()), False,
+                   SubstructureRedirectMask | SubstructureNotifyMask, &xev);
+        return;
+    }
+#endif
+#ifdef Q_WS_WIN
+    if (tlw->isWindow() && !tlw->testAttribute(Qt::WA_DontShowOnScreen) && !qt_widget_private(tlw)->hasHeightForWidth()) {
```

Comment:
Note to self: Calling the implementation in the base class was thought too naive and will not work!

However, I made a mistake and used Qt4's code by accident. I am currently replacing this code with recent code, where all of those nasty ifdef's vanished for good and code is cleaned up (For example `hasHeightForWidth` is now public member of `QWidget`) :+1:.

The whole topic relates to Qt's abstraction layer QPA (In Qt4 it was QWS). The only (sane) solution to this problem is actually by extending Qt itself (got no reply yet -> I'm going to repost).